src/fast-path.js

```diff
@@ -373,7 +373,7 @@ FastPath.prototype.needsParens = function(options) {
           parent.type === "GenericTypeAnnotation" ||
           parent.type === "TSTypeReference") &&
         (node.typeAnnotation.type === "TypeAnnotation" &&
-          node.typeAnnotation.typeAnnotation.type !== "TSFunctionType" &&
+          node.typeAnnotation.typeAnnotation.type !== "TypeAnnotation" &&
```

Comment:
Was the node type renamed at some point? If not, we might still need the TSFunctionType here

Reply:
oh.. it's my mistake.
This test is added by @azz to keep parens around function type. (https://github.com/prettier/prettier/commit/e8f12cf62149bf9e6456208c49b503371395ac99)

Are there any special cases that we should keep parens?

Reply:
```
type F = (() => string) | (() => boolean);
type F = (() => string) & (() => boolean);
```

Reply:
```ts
type F = () => string | number;
type F = (() => string) | number;
```

There may be more

Reply:
@azz, do you have any suggestion?

1. keep current behaviors?
2. strip parens as more as possible?
3. add parens around function type to make the output more consistent?

Reply:
I'd say go with parens only for intersection and union types. There may be more but we'll catch them eventually. We usually run Prettier on big TS projects before a release so we should catch any regressions.